RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitHeterogeneous.cc

```diff
@@ -155,22 +166,39 @@ void SiPixelRecHitHeterogeneous::acquireGPUCuda(const edm::HeterogeneousEvent& i
   edm::Handle<siPixelRawToClusterHeterogeneousProduct::GPUProduct> hinput;
   iEvent.getByToken<Input>(token_, hinput);
 
-  gpuAlgo_->makeHitsAsync(*hinput, fcpe->d_paramsOnGPU, cudaStream);
+  edm::Handle<reco::BeamSpot> bsHandle;
+  iEvent.getByToken( tBeamSpot, bsHandle);
+  float bs[3] = {0.f};
+  if(bsHandle.isValid()) {
+    const auto  & bsh = *bsHandle;
+    bs[0]=bsh.x0(); bs[1]=bsh.y0(); bs[2]=bsh.z0();
+  }
+
+
+  gpuAlgo_->makeHitsAsync(*hinput, bs, fcpe->d_paramsOnGPU, cudaStream);
 }
 
 void SiPixelRecHitHeterogeneous::produceGPUCuda(edm::HeterogeneousEvent& iEvent, const edm::EventSetup& iSetup, cuda::stream_t<>& cudaStream) {
   const auto& hits = gpuAlgo_->getOutput(cudaStream);
+  auto output = std::make_unique<GPUProduct>();
+  output->hits_d = hits.gpu_d;
 
-  // Need the CPU clusters to
+ // Need the CPU clusters to
   // - properly fill the output DetSetVector of hits
   // - to set up edm::Refs to the clusters
   edm::Handle<SiPixelClusterCollectionNew> hclusters;
   iEvent.getByToken(clusterToken_, hclusters);
 
-  auto output = std::make_unique<SiPixelRecHitCollectionNew>();
-  run(hclusters, *output, hits);
 
-  iEvent.put(std::move(output));
+  iEvent.put<Output>(std::move(output), [this, hits, hclusters](const GPUProduct&, CPUProduct& cpu) {
+      this->convertGPUtoCPU(hclusters, hits, cpu);
```

Comment:
You could avoid capturing `hits` by using the `HitsOnCPU` as the `GPUProduct`.

Reply:
what is the lifetime of the lambda? in principle in this way once the lambda is gone hits are gone as well
while as part of GPUProduct will stay in the event